database/pebble/db.go

```diff
@@ -0,0 +1,463 @@
+type batch struct {
+	batch *pebble.Batch
+	db    *Database
+	size  int
+
+	// Support batch rewrite
+	applied atomic.Bool
+}
+
+func (db *Database) NewBatch() database.Batch {
+	return &batch{
+		db:    db,
+		batch: db.db.NewBatch(),
+	}
+}
+
+func (b *batch) Put(key, value []byte) error {
+	b.size += len(key) + len(value) + pebbleByteOverHead
+	return b.batch.Set(key, value, pebble.NoSync)
+}
+
+func (b *batch) Delete(key []byte) error {
+	b.size += len(key) + pebbleByteOverHead
+	return b.batch.Delete(key, pebble.NoSync)
+}
+
+func (b *batch) Size() int { return b.size }
+
+func (b *batch) Write() error {
+	// write causes panic if the db has already been closed
+	if b.db.closed.Get() {
+		return database.ErrClosed
+	}
+
+	// Support batch rewrite
+	// The underlying pebble db doesn't support batch rewrites but panics instead
+	// We have to create a new batch which is a kind of duplicate of the given
+	// batch(arg b) and commit this new batch on behalf of the given batch.
+	if b.applied.Load() {
+		// the given batch b has already been committed
+		// Don't Commit it again, got panic otherwise
+		// Create a new batch to do Commit
+		newbatch := &batch{
+			db:    b.db,
+			batch: b.db.db.NewBatch(),
+		}
+
+		// duplicate b.batch to newbatch.batch
+		if err := newbatch.batch.Apply(b.batch, nil); err != nil {
+			return err
+		}
+		return updateError(newbatch.batch.Commit(pebble.NoSync))
+	}
+	// mark it for alerady committed
+	b.applied.Store(true)
+
+	return updateError(b.batch.Commit(pebble.NoSync))
+}
+
+func (b *batch) Reset() {
+	b.batch.Reset()
+	b.size = 0
+}
+
+func (b *batch) Replay(w database.KeyValueWriterDeleter) error {
+	reader := b.batch.Reader()
+	for {
+		kind, k, v, ok := reader.Next()
+		if !ok {
+			return nil
+		}
+		switch kind {
+		case pebble.InternalKeyKindSet:
+			if err := w.Put(k, v); err != nil {
+				return err
+			}
+		case pebble.InternalKeyKindDelete:
+			if err := w.Delete(k); err != nil {
+				return err
+			}
+		default:
+			return fmt.Errorf("%w: %v", ErrInvalidOperation, kind)
+		}
+	}
+}
+
+func (b *batch) Inner() database.Batch { return b }
+
+type iter struct {
+	db       *Database
+	iter     *pebble.Iterator
+	setFirst bool
+
+	valid bool
+	err   error
+}
+
+func (db *Database) NewIterator() database.Iterator {
+	// Don't call NewIter of pebble after the db closed. It panics otherwise.
+	if db.closed.Get() {
+		return &iter{
+			db: db,
+		}
+	}
+
+	return &iter{
+		db:   db,
+		iter: db.db.NewIter(&pebble.IterOptions{}),
+	}
+}
+
+func (db *Database) NewIteratorWithStart(start []byte) database.Iterator {
+	if db.closed.Get() {
+		return &iter{
+			db: db,
+		}
+	}
+
+	return &iter{
+		db:   db,
+		iter: db.db.NewIter(&pebble.IterOptions{LowerBound: start}),
+	}
+}
+
+// bytesPrefix returns key range that satisfy the given prefix.
+// This only applicable for the standard 'bytes comparer'.
+func bytesPrefix(prefix []byte) *pebble.IterOptions {
```

Comment:
I feel like instead of returning `pebble.IterOptions` here we could just return `prefix` and `limit` separately and let the caller create the config.